src/aggregated/circulating_supply/circulating_supply_provider.rs

```diff
@@ -0,0 +1,182 @@
+use std::ops::{Add, Div};
+use std::str::FromStr;
+use std::time::{Duration, SystemTime};
+
+use actix::Addr;
+use actix_diesel::Database;
+use actix_web::rt::time;
+use bigdecimal::{BigDecimal, ToPrimitive};
+use diesel::PgConnection;
+use tokio::time::Instant;
+use tracing::{error, info};
+
+use crate::aggregated::account_details;
+use crate::aggregated::account_details::get_account_balance;
+use crate::aggregated::circulating_supply::lockup;
+use crate::aggregated::circulating_supply::lockup::TRANSFERS_ENABLED;
+use crate::db_adapters::accounts;
+use crate::db_adapters::aggregated::circulating_supply::{
+    add_circulating_supply, get_precomputed_circulating_supply_for_timestamp,
+};
+use crate::db_adapters::blocks::get_latest_block_before_timestamp;
+use crate::models::aggregated::circulating_supply::CirculatingSupply;
+use crate::models::Block;
+
+const DAY: std::time::Duration = std::time::Duration::from_secs(86400);
+
+// Compute circulating supply on a daily basis, starting from 13 Oct 2020
+// (Transfers enabled moment on the Mainnet), and put it to the Indexer DB.
+// Circulating supply is calculated by the formula:
+// total_supply - sum(locked_tokens_on_each_lockup) - sum(locked_foundation_account)
+// The value is always computed for the last block in a day (UTC).
+pub(crate) async fn compute_circulating_supply(
+    view_client: Addr<near_client::ViewClientActor>,
+    pool: Database<PgConnection>,
+) {
+    let retry_duration = DAY.div(12);
+
+    // Adding 10 minutes to be sure that the picture is finalized and we can collect all needed data
+    let ten_minutes: u64 = 10 * 60 * 1000_000_000;
+    let mut day_to_compute =
+        TRANSFERS_ENABLED - TRANSFERS_ENABLED % (DAY.as_nanos() as u64) + ten_minutes;
+
+    loop {
+        time::sleep_until(get_instant_for_timestamp(day_to_compute)).await;
```

Comment:
I find this super odd and it took me a few brain-cycles to reason about. `get_instant_for_timestamp` looks suspicious at the first glance.

This whole function tries to be too clever by dealing with back-filling and daily updates in a single place. I don't have immediate suggestions, but I wanted to share the feedback from a person who reads it the first time.

Reply:
Thank you
While I was explaining why it's the best solution that we could have here, I created a simpler one. 😄